### PR TITLE
Cleanup ConfigurationRole to remove unnecessary functionality

### PR DESCRIPTION
**Pull request by tresat**
This class provides features no longer needed, this PR removes these:
- Remove pre-defined deprecated roles - these can now be replaced with migration roles.
- Remove deprecated fields from configuration roles.
- Remove ability to create custom role.

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationRole.java [hunk 1]
```diff
@@ -141,74 +102,4 @@ private static String describeDeprecation(boolean deprecated) {
             return deprecated ? " " + IS_DEPRECATED : "";
         }
     }
-
-    /**
-     * A custom implementation of {@link ConfigurationRole} that allows for non-standard combinations of usage characteristics and deprecations.
-     */
-    final class CustomConfigurationRole implements ConfigurationRole {
```
Comment (jvandort) [inline subprojects/core/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationRole.java:148]: Seems like this class could still be useful? We create anonymous implementations twice. Once in prod code and once in tests. If we pull this out to a top-level `DefaultConfigurationRole` we could reduce duplication.

**Review by jvandort — APPROVED** [review]

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationRole.java (hunk 1 above)
Comment (tresat) [reply subprojects/core/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationRole.java:148]: I don't really think the usage in the test should be counted towards the total - there's plenty of dummy implementations in test code.

The usage in prod code is also part of the copy functionality, which I'm going to explore deprecating next.  So I don't think this is worth doing.

**Comment by tresat** [comment]
@bot-gradle test and merge

**Comment by tresat** [comment]
@bot-gradle test and merge

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationRole.java (hunk 1 above)
Comment (tresat) [reply subprojects/core/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationRole.java:148]: Bah.  I say this and then a very minor regression fails the merge, which, looking at this change, might be from creating that anonymous inner type.  Going to try this and see if it helps.

**Pull request merged**